pmlc/dialect/pxa/transforms/stencil_xsmm.cc

```diff
@@ -0,0 +1,176 @@
+// Copyright 2020 Intel Corporation
+
+#include "pmlc/dialect/pxa/transforms/stencil_generic.h"
+
+// TODO: Just seeing if the stencil.cc includes work
+#include "mlir/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassOptions.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+
+#include "pmlc/util/logging.h"
+#include "pmlc/util/util.h"
+
+// TODO: includes etc
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+
+std::vector<size_t> TODOTempPowerOfTwoGenerator(size_t range) {
+  std::vector<size_t> out;
+  for (size_t r = 1; r <= range; r *= 2) {
+    out.push_back(r);
+  }
+  return out;
+}
+
+} // namespace
+
+class StencilXSMM : public StencilGeneric {
+private:
+  llvm::Optional<LoadStoreOps> capture() {
+    // Looking for load..load..mul..reduce..terminator
+    LoadStoreOps ret;
+    const unsigned kNumValidInstrInGemmRegion = 5;
+    auto *body = op.getBody();
+
+    // Verify the number of ops
+    if (body->getOperations().size() != kNumValidInstrInGemmRegion) {
+      IVLOG(5, "The AffineParallelOp region didn't have the right number of "
+               "instructions for a GEMM");
+      return llvm::Optional<LoadStoreOps>(); // i.e. fail to pattern-match
+    }
+
```

Comment:
I think it's possible to use `m_Op` or a variant of it that uses `detail::RecursivePatternMatcher` from https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/IR/Matchers.h#L263.  I think this would make the pattern matching and capturing a single line of code.

Reply:
Thanks, I'll take a look